File: libraries/python/src/pandaprobe/wrappers/mistral/wrapper.py
"""wrap_mistral — automatic LLM span instrumentation for the Mistral AI client."""

from __future__ import annotations

import functools
import logging
from typing import Any, TypeVar

from pandaprobe.wrappers._base import (
    AsyncStreamReducer,
    SyncStreamReducer,
    close_llm_span,
    error_llm_span,
    safe_serialize,
)
from pandaprobe.wrappers.mistral.utils import enter_mistral_span, strip_unset

logger = logging.getLogger("pandaprobe")

T = TypeVar("T")


def wrap_mistral(client: T) -> T:
    """Instrument a ``mistralai.Mistral`` client for automatic LLM span tracing.

    Monkey-patches ``chat.complete`` / ``chat.complete_async`` (blocking) and
    ``chat.stream`` / ``chat.stream_async`` (streaming) so every call
    automatically produces an LLM span.

    Returns the same client instance (mutated).
    """
    _patch_chat_complete(client)
    _patch_chat_complete_async(client)
    _patch_chat_stream(client)
    _patch_chat_stream_async(client)
    return client


# ---------------------------------------------------------------------------
# Patching helpers
# ---------------------------------------------------------------------------


def _patch_chat_complete(client: Any) -> None:
    if not hasattr(client, "chat") or not hasattr(client.chat, "complete"):
        return
    original = client.chat.complete
    client.chat.complete = _sync_blocking_wrapper(original)


def _patch_chat_complete_async(client: Any) -> None:
    if not hasattr(client, "chat") or not hasattr(client.chat, "complete_async"):
        return
    original = client.chat.complete_async
    client.chat.complete_async = _async_blocking_wrapper(original)


def _patch_chat_stream(client: Any) -> None:
    if not hasattr(client, "chat") or not hasattr(client.chat, "stream"):
        return
    original = client.chat.stream
    client.chat.stream = _sync_stream_wrapper(original)


def _patch_chat_stream_async(client: Any) -> None:
    if not hasattr(client, "chat") or not hasattr(client.chat, "stream_async"):
        return
    original = client.chat.stream_async
    client.chat.stream_async = _async_stream_wrapper(original)


# ---------------------------------------------------------------------------
# Sync / async blocking wrappers
# ---------------------------------------------------------------------------


def _sync_blocking_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = strip_unset(kwargs)
        span_ctx = enter_mistral_span(cleaned, "mistral-chat")
        try:
            response = original(*args, **kwargs)
            _finish_mistral_span(span_ctx, response)
            return response
        except Exception as exc:
            error_llm_span(span_ctx, exc)
            raise

    return wrapper


def _async_blocking_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    async def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = strip_unset(kwargs)
        span_ctx = enter_mistral_span(cleaned, "mistral-chat")
        try:
            response = await original(*args, **kwargs)
            _finish_mistral_span(span_ctx, response)
            return response
        except Exception as exc:
            error_llm_span(span_ctx, exc)
            raise

    return wrapper


# ---------------------------------------------------------------------------
# Sync / async streaming wrappers
# ---------------------------------------------------------------------------


def _sync_stream_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = strip_unset(kwargs)
        span_ctx = enter_mistral_span(cleaned, "mistral-chat")
        try:
            stream = original(*args, **kwargs)
            return _MistralSyncStream(stream, span_ctx)
        except Exception as exc:
            error_llm_span(span_ctx, exc)
            raise

    return wrapper


def _async_stream_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    async def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = strip_unset(kwargs)
        span_ctx = enter_mistral_span(cleaned, "mistral-chat")
        try:
            stream = await original(*args, **kwargs)
            return _MistralAsyncStream(stream, span_ctx)
        except Exception as exc:
            error_llm_span(span_ctx, exc)
            raise

    return wrapper


# ---------------------------------------------------------------------------
# Mistral-specific stream reducers
# ---------------------------------------------------------------------------


class _MistralSyncStream(SyncStreamReducer):
    def reduce_chunks(self, span_ctx: Any, chunks: list[Any]) -> None:
        _reduce_mistral_stream(span_ctx, chunks)


class _MistralAsyncStream(AsyncStreamReducer):
    def reduce_chunks(self, span_ctx: Any, chunks: list[Any]) -> None:
        _reduce_mistral_stream(span_ctx, chunks)


def _reduce_mistral_stream(span_ctx: Any, chunks: list[Any]) -> None:
    """Reduce Mistral streaming events into final span data.

    Each event yielded by ``client.chat.stream(...)`` exposes a ``.data``
    attribute holding the actual chunk (a ``CompletionEvent``).  The chunk's
    ``choices[0].delta.content`` carries the text delta and the terminal
    chunk includes ``usage``.
    """
    content_parts: list[str] = []
    model: str | None = None
    usage: dict[str, int] = {}

    for event in chunks:
        chunk = getattr(event, "data", event)

        chunk_model = getattr(chunk, "model", None)
        if chunk_model:
            model = chunk_model

        choices = getattr(chunk, "choices", None) or []
        if choices:
            delta = getattr(choices[0], "delta", None)
            if delta is not None:
                text = getattr(delta, "content", None)
                if isinstance(text, str) and text:
                    content_parts.append(text)

        chunk_usage = getattr(chunk, "usage", None)
        if chunk_usage is not None:
            usage = _extract_mistral_usage(chunk_usage) or usage

    if content_parts:
        span_ctx.set_output(
            {"messages": [{"role": "assistant", "content": "".join(content_parts)}]},
        )
    if model:
        span_ctx.set_model(model)
    if usage:
        span_ctx.set_token_usage(**usage)

    close_llm_span(span_ctx)


# ---------------------------------------------------------------------------
# Response extraction helpers
# ---------------------------------------------------------------------------


def _finish_mistral_span(span_ctx: Any, response: Any) -> None:
    """Extract output, model, and token usage from a Mistral chat response."""
    if span_ctx is None:
        return

    try:
        choices = getattr(response, "choices", None) or []
        if choices:
            message = getattr(choices[0], "message", None)
            if message is not None:
                serialized = safe_serialize(message)
                if isinstance(serialized, dict):
                    serialized.setdefault("role", "assistant")
                    span_ctx.set_output({"messages": [serialized]})
                else:
                    span_ctx.set_output({"messages": [{"role": "assistant", "content": serialized}]})
    except Exception as exc:
        logger.debug("Error extracting Mistral response output: %s", exc)

    try:
        model = getattr(response, "model", None)
        if model:
            span_ctx.set_model(model)
    except Exception as exc:
        logger.debug("Error extracting Mistral response model: %s", exc)

    try:
        usage = getattr(response, "usage", None)
        if usage is not None:
            token_kwargs = _extract_mistral_usage(usage)
            if token_kwargs:
                span_ctx.set_token_usage(**token_kwargs)
    except Exception as exc:
        logger.debug("Error extracting Mistral response usage: %s", exc)

    close_llm_span(span_ctx)


def _extract_mistral_usage(usage: Any) -> dict[str, int]:
    """Map Mistral usage fields onto canonical PandaProbe token usage names.

    Mistral already uses ``prompt_tokens`` / ``completion_tokens`` /
    ``total_tokens`` natively so this is mostly a defensive coercion to
    integers.
    """
    prompt = getattr(usage, "prompt_tokens", 0) or 0
    completion = getattr(usage, "completion_tokens", 0) or 0
    total = getattr(usage, "total_tokens", 0) or 0

    out: dict[str, int] = {}
    if isinstance(prompt, int) and prompt > 0:
        out["prompt_tokens"] = prompt
    if isinstance(completion, int) and completion > 0:
        out["completion_tokens"] = completion
    if isinstance(total, int) and total > 0:
        out["total_tokens"] = total
    return out

File: libraries/typescript/src/wrappers/mistral/wrapper.ts
/** wrapMistral — automatic LLM span instrumentation for the Mistral AI JS client. */

import type { SpanContext } from "../../tracing/span.js";
import { closeLlmSpan, errorLlmSpan, safeSerialize, tokenUsage, wrapAsyncStream } from "../base.js";
import { enterMistralSpan } from "./utils.js";

type Any = any;

/**
 * Instrument a `Mistral` client.
 *
 * Patches `chat.complete` (blocking) and `chat.stream` (streaming — events
 * expose the chunk on `.data`). Returns the same client instance (mutated).
 */
export function wrapMistral<T>(client: T): T {
  patchComplete(client as Any);
  patchStream(client as Any);
  return client;
}

function patchComplete(client: Any): void {
  const chat = client?.chat;
  if (!chat || typeof chat.complete !== "function") {
    return;
  }
  const original = chat.complete.bind(chat);
  chat.complete = async (params: Any, ...rest: Any[]): Promise<Any> => {
    const span = enterMistralSpan(params ?? {}, "mistral-chat");
    try {
      const response = await original(params, ...rest);
      finishMistralSpan(span, response);
      return response;
    } catch (exc) {
      errorLlmSpan(span, exc);
      throw exc;
    }
  };
}

function patchStream(client: Any): void {
  const chat = client?.chat;
  if (!chat || typeof chat.stream !== "function") {
    return;
  }
  const original = chat.stream.bind(chat);
  chat.stream = async (params: Any, ...rest: Any[]): Promise<Any> => {
    const span = enterMistralSpan(params ?? {}, "mistral-chat");
    try {
      const stream = await original(params, ...rest);
      return wrapAsyncStream(stream as AsyncIterable<Any>, span, reduceMistralStream);
    } catch (exc) {
      errorLlmSpan(span, exc);
      throw exc;
    }
  };
}

function extractMistralUsage(usage: Any): Record<string, number> | null {
  if (!usage) {
    return null;
  }
  return tokenUsage({
    prompt_tokens: usage.promptTokens ?? usage.prompt_tokens,
    completion_tokens: usage.completionTokens ?? usage.completion_tokens,
    total_tokens: usage.totalTokens ?? usage.total_tokens,
  });
}

function setUsage(span: SpanContext, usage: Record<string, number> | null): void {
  if (!usage) {
    return;
  }
  const { prompt_tokens, completion_tokens, ...extra } = usage;
  span.setTokenUsage({ promptTokens: prompt_tokens ?? 0, completionTokens: completion_tokens ?? 0, ...extra });
}

function reduceMistralStream(span: SpanContext, events: Any[]): void {
  const contentParts: string[] = [];
  let model: string | undefined;
  let usage: Record<string, number> | null = null;

  for (const event of events) {
    const chunk = event?.data ?? event;
    if (chunk?.model) {
      model = chunk.model;
    }
    const delta = chunk?.choices?.[0]?.delta;
    if (delta && typeof delta.content === "string" && delta.content) {
      contentParts.push(delta.content);
    }
    if (chunk?.usage) {
      usage = extractMistralUsage(chunk.usage) ?? usage;
    }
  }

  try {
    if (contentParts.length > 0) {
      span.setOutput({ messages: [{ role: "assistant", content: contentParts.join("") }] });
    }
    if (model) {
      span.setModel(model);
    }
    setUsage(span, usage);
  } catch {
    // best-effort
  }
  closeLlmSpan(span);
}

function finishMistralSpan(span: SpanContext | null, response: Any): void {
  if (span === null) {
    return;
  }
  try {
    const message = response?.choices?.[0]?.message;
    if (message != null) {
      const serialized = safeSerialize(message);
      if (serialized && typeof serialized === "object" && !Array.isArray(serialized)) {
        const copy = { role: "assistant", ...(serialized as Record<string, unknown>) };
        span.setOutput({ messages: [copy] });
      } else {
        span.setOutput({ messages: [{ role: "assistant", content: serialized }] });
      }
    }
  } catch {
    // ignore
  }

  try {
    if (response?.model) {
      span.setModel(response.model);
    }
  } catch {
    // ignore
  }

  try {
    setUsage(span, extractMistralUsage(response?.usage));
  } catch {
    // ignore
  }

  closeLlmSpan(span);
}
